### PR TITLE
[Datasets] Allow for len(Dataset)

### LINKED ISSUE #25147: [Datasets] Add `Dataset.__len__`
### Description

Implement the `Dataset.__len__` method.

### Use case

Would look slightly nicer for some computations

Instead of 

```
scores.sum(on="correct") / scores.count()
```

we could do

```
scores.sum(on="correct") / len(scores)
```

### PR DESCRIPTION
**Pull request by bushshrub**
## Why are these changes needed?

Small QOL change that allows for `len(Dataset)` to be used rather than calling `Dataset.count()`

## Related issue number

Closes #25147 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by bveeramani — APPROVED** [review]
Hey @xiurobert, thanks for the contribution! This looks good to me.

**Review by ericl — CHANGES_REQUESTED** [review]
Typically users expect len to be fast. However, count can execute a full distributed job. How about raising an error message instead telling users to use `count()`?

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -3424,6 +3424,9 @@ def __repr__(self) -> str:
     def __str__(self) -> str:
         return repr(self)
 
+    def __len__(self) -> int:
+        return self.count()
```
Comment (bveeramani) [inline python/ray/data/dataset.py:3428]: Actually, could you add a simple test to `test_dataset`?

**Review by bveeramani — CHANGES_REQUESTED** [review]

**Comment by xiurobert** [comment]
I can do that, yeah.

File: python/ray/data/dataset.py (hunk 1 above)
Comment (xiurobert) [reply python/ray/data/dataset.py:3428]: Sure.

Comment (xiurobert) [reply python/ray/data/dataset.py:3428]: https://github.com/ray-project/ray/pull/25152/commits/41e2dfbe8a1cd7d35edd08b7a07adbf3880d9085

**Comment by xiurobert** [comment]
> Typically users expect len to be fast. However, count can execute a full distributed job. How about raising an error message instead telling users to use `count()`?

What about throwing them a warning that this can be slow, and executing it anyway?

**Comment by ericl** [comment]
The problem with the warning is you have code printing lots of warnings: better to have an error and the user fix the path to use count().

**Comment by xiurobert** [comment]
Currently `len(Dataset)` already returns an error. Should we raise a different exception instead telling users to use `ds.count()`?

**Comment by xiurobert** [comment]
Is there a specific exception that is implemented within the project for such actions? Or should one be implemented instead?

**Comment by ericl** [comment]
I think AttributeError (or whatever the original type), with a better error message would be the most natural.

**Comment by ericl** [comment]
Huh some very strange failures. It seems that `not ds` is calling `__len__`:
```
E                         File "/ray/python/ray/data/dataset_pipeline.py", line 315, in __next__
  | E                           if not ds:
  | E                         File "/ray/python/ray/data/dataset.py", line 3429, in __len__
  | E                           "Use `ds.count()` to compute the length of a distributed Dataset. "
```

**Comment by xiurobert** [comment]
Seems as if Python is using `__len__` internally to check for the `ds` being `None`.

**Comment by jianoaix** [comment]
+1 to use count() since this would be a non-trivial distributed operation.
If implementing __len__ and raising error would create other errors, can we just leave it as it is?

**Comment by ericl** [comment]
Seems we should implement `__bool__` to always return True. Python won't check `__len__` if bool is defined.

**Comment by xiurobert** [comment]
What about __bool__ returning whether the block_count > 0?

**Comment by ericl** [comment]
I think that's probably too risky (breaks backwards compatibility to suddenly be False now).

**Comment by xiurobert** [comment]
Ok then, probably best to set it up to just be True all the time.

**Review by ericl — APPROVED** [review]

**Comment by clarkzinzow** [comment]
A seemingly unrelated test is failing: https://buildkite.com/ray-project/ray-builders-pr/builds/33563#01810385-e98b-4595-af19-0cdaee7bd0ff

**Comment by xiurobert** [comment]
Oh, I accidentally set `__bool__` to always be false, haha. Could be that. I'll patch it asap.

**Comment by xiurobert** [comment]
Let me clean up this branch, I'll do a rebase.

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -3414,6 +3414,15 @@ def __repr__(self) -> str:
     def __str__(self) -> str:
         return repr(self)
 
+    def __bool__(self) -> bool:
+        return True
```
Comment (bveeramani) [inline python/ray/data/dataset.py:3418]: Could you add a comment here? I don't think readers would understand why we implement `__bool__` without reading this thread.

**Review by bveeramani — APPROVED** [review]
Hey, thanks for your work on this! Looks good to me.

If it's not too much trouble, I had a minor suggestion.

File: python/ray/data/dataset.py (hunk 2 above)
Comment (xiurobert) [reply python/ray/data/dataset.py:3418]: Yup, seems like a good idea too.

**Comment by clarkzinzow** [comment]
Lint is failing: https://buildkite.com/ray-project/ray-builders-pr/builds/34248#01812abb-7d08-4658-9731-4606310a92f9

Could you run `ci/lint/format.sh` and push up the changes?

**Comment by xiurobert** [comment]
All done @clarkzinzow

**Review by clarkzinzow — APPROVED** [review]
LGTM, thank you for the contribution @xiurobert! 🙌

**Pull request merged**